Update/maintain build test

cevich opened the pull request:
I got a little carried away working on this test today, decided to basically re-write most of the test since it was not very modular at all.  The bad news is, I ran out of time working on the postprocessing stuff.

What I did was to define a fixed structure for holding the build information, instead of a loose dictionary that wasn't well represented.  Then I consolidated all the config. options to fit the structure, and stuck it in place for ``initialize()`` and ``run_once()``.  Now it's all nice, clean, and well defined what each thing is.  Though untested (they're may be typos).

Finally, I worked my way down and re-wrote the postprocessing methods to be much more general.  I left off at ``postprocess_created_containers()``.  I think ``postprocess_output()`` can probably just be deleted, since those checks were consolidated into ``postprocess_positive()`` and ``postprocess_negative()``.  My thinking was, make the base-class ``postprocess()`` method abstract (or just remove it), then override it and ``cleanup()`` inside each different sub-subtest class.  That way they have exact control over which checks are performed and how they treat the meaning of ``positive_build_regex`` or ``positive_build_regex``.

Anyway, take it for what it is.  If you want to continue, take some of these ideas, or just forget my work and try and fix the existing test, either way is fine.  I probably should have looked closer at this test before saying 'okay' to fix it, I'm sorry I didn't realize how much of a mess it was :confounded:  (so, totally not your fault).  In any case, however you go forward is fine.  I see this test along with many others are in need of some cleaning up, so learn and have fun.  I'll be back in a few weeks, cheers!

cevich commented:
Okay, more complete now, I've noted all my fixes/changes in commit message, and I've pylint'ed everything.  I think I fixed the cache problem you found, and many other problems I found (it was ugly!).  It's not perfect, but I think much better than before, and should work nicely with the 'FROM' updating you've put in.  Still it's all untested, so that's next :smile:

luwensu commented on subtests/docker_cli/build/build.py:
```diff
@@ -79,6 +87,8 @@ def setup(self):
 class build_base(subtest.SubSubtest):
 
     def dockerfile_dir_path(self, dir_path):
+        if isinstance(dir_path, basestring) or dir_path is None:
```
The isinstance() will return true normally (I'm not try with other value, just dir_path = "/full"), is it excepted for some reseaons? Or should we change to "not isinstance()" to avoid go into this if....

luwensu commented on subtests/docker_cli/build/build.py:
```diff
@@ -114,213 +124,191 @@ def dockerfile_repo_replace(self, path, repo):
+        if build_def.positive_build_regex != '':
+            positive_regex = re.compile(build_def.positive_build_regex)
+            for output in (build_def.dockercmd.stdout,
+                           build_def.dockercmd.stderr):
```
The build_def.dockercmd.stderr is going to be empty if the case run successfully, it pass after i add a 
if output:
   mobj = positive_regex.search(output)

cevich replied:
Yeah, you're right, but it's worse.  This code `postprocess_positive()`, etc. really stinks, I hate it.  It requires configuration to be tightly-coupled with the sub-subtest `postprocess()` code (i.e. so what is purpose of configuration?).

I think the 'strategy' or 'command' design pattern, would work well here.  The idea is to have a configuration string that lists the post-process steps needed.  Then in the test's `postprocess()` we parse the string and call the methods asked for.  Though as always, "the devil is in the details": writing a parser is almost always a PITA :cry: I'll see what I can do...

cevich replied:
...okay, fixed this particular method, so it takes a parameter to help decide on checking stdout or stderr...

luwensu commented on subtests/docker_cli/build/build.py:
```diff
@@ -114,213 +124,191 @@ def dockerfile_repo_replace(self, path, repo):
+        negative_build_regex = source['negative_build_regex'].strip()
+        base_repo_fqin = DockerImage.full_name_from_defaults(source)
+        use_config_repo = source['use_config_repo']
+        dockerfile_dir_path = source['dockerfile_dir_path'].strip()
```
Looks like dockerfile_dir_path lack of the system path which get from  self.parent_subtest.srcdir from func def dockerfile_dir_path, it's /full(for example) in my env now....

cevich replied:
Ahh, woops, did I forget to add that back in?  I'll check it out, thanks...

cevich replied:
...okay, yep, my mistake, fixed now.

luwensu commented on subtests/docker_cli/build/build.py:
```diff
@@ -114,213 +124,191 @@ def dockerfile_repo_replace(self, path, repo):
+
+        img_names = [di.full_name for di in images]
+        imf_ids = [di.short_id for di in images]
+        self.failif(build_def.image_name not in img_names,
```
the images[0] is something like the whole string
"full_name:local_path_5ku7:latest LONG_ID:66c46abde1df0238e19540c7450151cfbe12d9d004d16acb08fa5aa1a4d4c6e2 CREATED:5 seconds ago SIZE:148.2 MB
" 
, so do we need a regular exp to match it just like what _postprocess_output does?

cevich replied:
Nope, what you're seeing is just the `__str__()` method for `DockerImage` instance.  You can check the attributes directly, so e.g. `di.short_id` will always be the 12-character "short" ID of the image.  What happens above, is each `full_name` attribute from every item in `images` (`DockerImage` instance list) is extracted into it's own list (`img_names`).  Then that makes it easy to verify if expected `build_def.image_name` is contained in list of all images or not.

luwensu commented on subtests/docker_cli/build/build.py:
```diff
@@ -114,213 +124,191 @@ def dockerfile_repo_replace(self, path, repo):
+
+        stdout = build_def.dockercmd.stdout
+        created_ids = [mobj.group(1)
+                       for mobj in RE_IMAGES.search(stdout)
```
Something goes wrong here..debugging
  File "/root/autotest/client/tests/build/subtests/docker_cli/build/build.py", line 288, in postprocess_created_images
    for mobj in RE_IMAGES.search(stdout)
TypeError: '_sre.SRE_Match' object is not iterable

cevich replied:
Ahh, okay, right the match-object returned by search cannot be looped over.  I think what we want here is `finditer` instead of `search`, thanks I'll check it out...

luwensu commented:
Ahhh...I don't want to see my code comment in this page....and sorry for i took a leave yesterday btw
I did a play around and find some issue that need confirm with you are above, i'd like to know if the structure need to be changed or just few small fixes are ok . And not finish it yet anyway...
If the structure is stable enough (it gives me some spirit that i'd like to follow in other case :) ), i'll merge and push it after fix and test.

cevich commented on subtests/docker_cli/build/build.py:
```diff
@@ -64,212 +72,243 @@ def setup(self):
-    def dockerfile_path(self, path):
-        if path[0] == '/':
+    def dockerfile_dir_path(self, dir_path):
+        if isinstance(dir_path, basestring) or dir_path is None:
```
> The isinstance() will return true normally (I'm not try with other value, just dir_path = "/full"), is it excepted for some 
> reseaons? Or should we change to "not isinstance()" to avoid go into this if....

Woops, you're correct, it should be `not isinstance(...)`, I'll fix it...

cevich commented:
> Ahhh...I don't want to see my code comment in this page....and sorry for i took a leave yesterday

No problem, thanks for taking the time to find my mistakes, clearly I could not see them myself :grinning: 

> If the structure is stable enough

Mostly, I think I do want to hack in some kind of command-pattern for post-process because there are just too many ways to do build and account for all the checking for every test in always the right way.

In my mind, most of these tests are in need of maintenance / updating.  So I feel need to put in your fixes and fix rest of test to make it right.  This is just stuck in my mind though, do not feel pressure for you fix everything yourself, small fixes/updates are fine and helpful also :grinning:

Okay, I will hack on this some more, I think it's just `postprocess_*()` that needs some final help...

cevich commented:
New commits:  Fixed your commit message (must have space between first line and rest of message).  Also fixed problems you found above.  Still working on adding command-pattern for `postprocess()`.

cevich commented:
.. Eek! okay, as expected, updating this to use the command-pattern for post-processing is going to be more involved than I envisioned.  Let's get _this_ original work, plus my updates working and merged, and I'll work on the post-processing update in parallel.  Would you mind checking what's here so-far and see if you spot any more problems?  Take your time.  Thanks!

cevich commented:
Pushed new commits:  Rebased off of latest master, renamed my commit to just apply ontop of yours, added a blank-line after your commit description (that's why it says there are merge conflicts).

luwensu commented:
okay, i merged this after resolved the conflict by hand (Some edit conflict between my origin branch and these commits), in case it will not block more future work as it includes lots of changes.
I made a mistake today, i used the branch before you fixed to test :(, good news is it passed after i did a small fix which i saw in your latest commit (except "cache" case, maybe because some options in the config file, not sure, working on it).
For something others need change like typos error, i'd like to push a new commit into the main project, thus it will not block the functional patches merged.

luwensu commented on subtests/docker_cli/build/build.py:
```diff
@@ -302,13 +347,40 @@ class git_path(build_base):
+
+    def make_builds(self, source):
+        first = source.deepcopy()
+        second = source.deepcopy()
```
Not sure here, it failed to execute in my RHEL7 env, does it related the python version or something else ?
It works through this way:
import copy
first = copy.deepcopy(source)
otherwise, it will pop
'dict' object has no attribute 'deepcopy'

cevich replied:
Nope, this is a typo on my part, you are correct `copy.deepcopy()` should be used.

cevich commented:
> For something others need change like typos error, i'd like to push a new commit into the main project, thus it will not block the functional patches merged.

Yep, no problem, just open up a new PR and stick it in.  For no-brain-required small commits, I'll just press "the green button" :smile:

cevich commented:
> except "cache" case

Ya, that one is going to be tricky b/c as near as I can tell, there is no interface for clearing or inspecting the cache, so generating new content all the time is the only option.  I'm fine if we need to just leave this as a kind of "smoke" test, and fine-tune it later.
